src/jltypes.c

```diff
@@ -523,6 +523,50 @@ JL_DLLEXPORT jl_value_t *jl_type_union(jl_value_t **ts, size_t n)
     return tu;
 }
 
+jl_value_t *simple_union(jl_value_t *a, jl_value_t *b)
+{
+    size_t nt = count_union_components(&a, 1);
+    nt += count_union_components(&b, 1);
+    jl_value_t **temp;
+    JL_GC_PUSHARGS(temp, nt+1);
+    size_t count = 0;
+    flatten_type_union(&a, 1, temp, &count);
+    flatten_type_union(&b, 1, temp, &count);
+    assert(count == nt);
+    size_t i, j;
+    for (i = 0; i < nt; i++) {
+        int has_free = 1; // TODO: is this worthwhile? temp[i] != NULL && jl_has_free_typevars(temp[i]);
+        for (j = 0; j < nt; j++) {
+            if (j != i && temp[i] && temp[j]) {
+                if (temp[i] == jl_bottom_type ||
+                    temp[j] == (jl_value_t*)jl_any_type ||
+                    jl_egal(temp[i], temp[j]) ||
+                    (!has_free && !jl_has_free_typevars(temp[j]) &&
+                     jl_subtype(temp[i], temp[j]))) {
```

Comment:
Perhaps we should also follows this rule here?
```julia
// issue #24521: don't merge Type{T} where typeof(T) varies
        !(jl_is_type_type(a) && jl_is_type_type(b) && jl_typeof(jl_tparam0(a)) != jl_typeof(jl_tparam0(b)))) {
```

Reply:
Good observation. I moved that here. I wanted to be careful not to make too big of a change at once, in case something in a package was depending on it significantly doing one thing. Maybe we should try to move more of that simple_tmeet function into simple_union, but I think a couple things are heuristic based there, so maybe not all of it should be combined.

Reply:
> but I think a couple things are heuristic based there, so maybe not all of it should be combined.

I think a basic rule is keeping the pointer identity if possible?
#49277 seems to be a good template and if we do that, I think we can move all type-specific branch into `simple_union` and let `simple_join` handle non-type bits along.

BTW, I didn't measure the performance influence.
But if we decided to turn this path on, then a good optimization to have is skipping all check if `temp[i]` and `temp[j]` come from the same part.
And perhaps this check could always be skipped for `omit_bad_union`?

Reply:
I was thinking of this check mostly, since I don't know if this is accurate for a Union in general:
```
    if (jl_is_typevar(a) && obviously_egal(b, ((jl_tvar_t*)a)->lb))
         return a;
```

That may be a good thing to check for during `jl_type_union` too to save repeated work in the case of repeatedly appending one element (e.g. `Union{Union{Union{A, B}, C}, Union{D, E}}`, since we only need to check the new elements against the old ones). A starting point for that might be to alloca a second buffer that holds the index of which where union starts from to fill the buffer. I think the trickiest ones to detect are cases like this, since "obviously" it can return the first object, but since we only do a subtype check (not equality), we may conclude the result should be that of prepending `A` to the second Union, and fail to realize that will give back the first Union:
```
Union{ Union{A,B,C}, Union{B,C} }
```
That suggests perhaps we should also do the same exhaustive sort of check afterwards too, as templated in https://github.com/JuliaLang/julia/pull/49277, to see if all the elements remaining are equivalent to an existing one. Once we merge this and https://github.com/JuliaLang/julia/pull/49277, I can try prototyping that additional check.

Reply:
> I was thinking of this check mostly, since I don't know if this is accurate for a Union in general:

I think it's accurate for `Union`, at least for those without free vars?
But since `Union` doesn't check it for now, I think we can leave these 2 branches in `simple_join`.

Reply:
> That suggests perhaps we should also do the same exhaustive sort of check afterwards too, as templated in https://github.com/JuliaLang/julia/pull/49277

Hoist some egal check here seems good?
Then we just need to count how many remaining `temp` comes from `a` or `b`.